test/ref/unpack_int4.cpp

```diff
@@ -0,0 +1,158 @@
+ * THE SOFTWARE.
+ */
+#include <migraphx/instruction.hpp>
+#include <migraphx/literal.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/register_target.hpp>
+#include <migraphx/verify.hpp>
+
+#include <test.hpp>
+
+TEST_CASE(unpack_int4)
+{
+    migraphx::program p;
+    auto* mm = p.get_main_module();
+    migraphx::shape s{migraphx::shape::uint8_type, {2, 1}};
+    auto l0 = mm->add_literal(migraphx::literal{s, {0xBA, 0xDC}});
+    mm->add_instruction(migraphx::make_op("unpack_int4"), l0);
+    p.compile(migraphx::make_target("ref"));
+    auto result = p.eval({}).back();
+    std::vector<uint8_t> results_vector(4);
+    result.visit([&](auto output) { results_vector.assign(output.begin(), output.end()); });
+    std::vector<uint8_t> gold{static_cast<uint8_t>(0x0A),
+                              static_cast<uint8_t>(0x0B),
+                              static_cast<uint8_t>(0x0C),
+                              static_cast<uint8_t>(0x0D)};
+    EXPECT(migraphx::verify::verify_rms_range(results_vector, gold));
+}
+
+TEST_CASE(unpack_int4_transposed)
+{
+    migraphx::program p;
+    auto* mm = p.get_main_module();
+    migraphx::shape s{migraphx::shape::uint8_type, {2, 2}, {1, 2}};
+    auto l0 = mm->add_literal(migraphx::literal{s, {0x1A, 0x2B, 0x3C, 0x4D}});
+    mm->add_instruction(migraphx::make_op("unpack_int4"), l0);
+    p.compile(migraphx::make_target("ref"));
+    auto result = p.eval({}).back();
+    std::vector<uint8_t> results_vector(8);
+    result.visit([&](auto output) { results_vector.assign(output.begin(), output.end()); });
+    std::vector<uint8_t> gold{static_cast<uint8_t>(0x0A),
+                              static_cast<uint8_t>(0x01),
+                              static_cast<uint8_t>(0x0B),
+                              static_cast<uint8_t>(0x02),
+                              static_cast<uint8_t>(0x0C),
+                              static_cast<uint8_t>(0x03),
+                              static_cast<uint8_t>(0x0D),
+                              static_cast<uint8_t>(0x04)};
+    EXPECT(migraphx::verify::verify_rms_range(results_vector, gold));
+}
+
+TEST_CASE(unpack_int4_broadcasted)
+{
+    migraphx::program p;
+    auto* mm = p.get_main_module();
+    migraphx::shape s{migraphx::shape::uint8_type, {4}, {1}};
+    auto l0  = mm->add_literal(migraphx::literal{s, {0x1A, 0x2B, 0x3C, 0x4D}});
+    auto l0b = mm->add_instruction(migraphx::make_op("multibroadcast", {{"out_lens", {4, 4}}}), l0);
+    mm->add_instruction(migraphx::make_op("unpack_int4"), l0b);
+    p.compile(migraphx::make_target("ref"));
+    auto result = p.eval({}).back();
+    std::vector<uint8_t> results_vector(32);
+    result.visit([&](auto output) { results_vector.assign(output.begin(), output.end()); });
+    std::vector<uint8_t> gold{
+        static_cast<uint8_t>(0x0A), static_cast<uint8_t>(0x01), static_cast<uint8_t>(0x0B),
```

Comment:
Unless there is some specific compiler issue, there should be no need for all the static_cast<uint8_t> for individual elements.. the vector type is already specified as uint8_t. There are many such instances..

Reply:
removed them